### PR TITLE
Add 3denvar option to run_jedi_exe.py

### LINKED ISSUE #402: Extend run_jedi_exe.py to launch 3denvar job
Currently `ush/run_jedi_exe.py` can launch fv3-jedi jobs for the following configurations:
- 3dhofx
- 4dhofx
- 3dvar
- letkf

A hybrid variational option is missing.  This issue is opened to extend `run_jedi_exe.py` to launch `fv3jedi_var.x` in 3denvar mode.

### PR DESCRIPTION
**Pull request by RussTreadon-NOAA**
The changes in this PR add the option to launch `fv3jedi_var.x` in 3denvar mode via `ush/run_jedi_exe.py`.   This functionality is useful since operational GSI-based global and regional DA systems run in a hybrid model in which the static background error is augmented with an ensemble component (envar).   

This PR adds four new files:
- `parm/atm/berror/hybvar_gsibec.yaml` - hybrid background error yaml file using `gsibec` for static-B along with a 10 member ensemble
- `test/atm/run_jedi_exe_3denvar.sh` - ctest using `run_jedi_exe.py` to launch `fv3jedi_var.x` in 3denvar mode 
- `ush/examples/run_jedi_exe/3denvar_hera.yaml` - configuration file used by `run_jedi_exe.py` to run 3denvar sample case on Hera
- `ush/examples/run_jedi_exe/3denvar_orion.yaml` - configuration file used by `run_jedi_exe.py` to run 3denvar sample case on Orion

Eleven existing files are modified:
- `CMakeLists.txt` - update gsibec to tag 1.0.7.   Previously pointed at head of gsibec develop.  This is problematic since the head of gsibec develop is not always function and/or requires specific hashes of other JEDI components.
- `parm/atm/obs/config/sondes.yaml` - replace `surface_geometric_height` with `surface_geopotential_height` to avoid `fv3jedi_var.x` segmentation fault
- `test/atm/CMakeLists.txt` - add `test_gdasapp_run_jedi_exe_3denvar`
- `test/genYAML_prep.sh` - replace `CASE_ENKF` with `CASE_ANL`
- `test/genYAML_prep_aero.sh` - replace `CASE_ENKF` with `CASE_ANL`
- `test/genYAML_prep_land.sh` - replace `CASE_ENKF` with `CASE_ANL`
- `ush/run_jedi_exe.py` - remove `atmos` from enkfgdas root path.   This change is necessitate by changes in the g-w GFS directory structure.   The `atmos` directory is now located in `memXXX` directories.
- `ush/ufsda/genYAML.py` - replace `CASE_ENKF` with `CASE_ANL`
- `ush/ufsda/misc_utils.py` - replace `CASE_ENKF` with `CASE_ANL`
- `ush/ufsda/stage.py` - add `atmos` to the `memXXX` path for ensemble members, enhance error checking for `dohybvar = .true.` option, add logic to ensure correct `gsibec` static-B is linked to the run directory.
- `ush/ufsda/ufs_yaml.py` - replace `CASE_ENKF` with `CASE_ANL`

Fixes #402

### CODE REVIEW

File: ush/ufsda/stage.py
Comment (kevindougherty-noaa) [inline ush/ufsda/stage.py]:
```
case_gsibec = case_anl if dohybvar else case
```

File: ush/ufsda/stage.py
Comment (kevindougherty-noaa) [inline ush/ufsda/stage.py]: Do you want print statements or should this be outputted to a logger file?

**Review by kevindougherty-noaa — COMMENTED** [review]
Just a couple of quick remarks. I will take a more in-depth look first thing Monday.

File: ush/ufsda/stage.py
Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Thanks.  I didn't realize the if-else structure could be collapsed into a single line.   The suggested line is succinct and readable.   

I made the change.   Subsequent testing revealed that the original logic is not correct.  `gsibec` requires that we use `case` independent of the value of `dohybvar`.   Neither the original logic nor suggested revision are necessary.   

The logical test was removed and `case` is used for all `gsibec` runs.  Change committed at 0ba33d8.

File: ush/ufsda/stage.py
Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Good suggestion.  `print` was replaced with `logging.info` at 0ba33d8.

Much to my surprise, no output is generated by `logging.info` when `run_jedi_exe.py`   I added a `print` statement before `logging.info` and only the `print` was echoed to stdout.

A check of `ush/run_jedi_exe.py` shows other instances of `logging.info` which do not generate printout when the script is executed.   This is odd because `run_jedi_exe.py` sets the logging level to `INFO`
```
def run_jedi_exe(yamlconfig):
    logging.basicConfig(format='%(asctime)s:%(levelname)s:%(message)s', level=logging.INFO, datefmt='%Y-%m-%d %H:%M:%S')
    # open YAML file to get config
    try:
        with open(yamlconfig, 'r') as yamlconfig_opened:
            all_config_dict = yaml.safe_load(yamlconfig_opened)
        logging.info(f'Loading configuration from {yamlconfig}')
```
When `run_jedi_exe.py` is executed, the `Loading configuration from ...` message is _not_ printed.   The only printout I see when executing `run_jedi_exe.py` is from R2D2.   I do not understand why we see this behavior.   Any ideas @CoryMartin-NOAA?

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: As a test, I commented out 
```
from ufsda.genYAML import genYAML
```
at the top of `run_jedi_exe.py`.    With this line commented out, `run_jedi_exe.py` generated info log messages up to the point where the script referenced `genYAML`.   

```
(gdasapp) Hera(hfe05):/scratch1/NCEPDEV/da/Russ.Treadon/git/global-workflow/test2/sorc/gdas.cd/ush$ ./run_jedi_exe.py -c examples/run_jedi_exe/3denvar_hera.yaml
2023-03-25 20:50:42:INFO:Loading configuration from examples/run_jedi_exe/3denvar_hera.yaml
2023-03-25 20:50:42:INFO:Created working directory /scratch2/NCEPDEV/stmp1/Russ.Treadon/gdas_single_test_3denvar.case_anl                              _c48
2023-03-25 20:50:44:INFO:Note: detected 80 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" en                              vironment variable.
2023-03-25 20:50:44:INFO:Note: NumExpr detected 80 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
2023-03-25 20:50:57:INFO:Using yamlconfig examples/run_jedi_exe/3denvar_hera.yaml
Traceback (most recent call last):
  File "./run_jedi_exe.py", line 177, in <module>
    run_jedi_exe(args.config)
  File "./run_jedi_exe.py", line 139, in run_jedi_exe
    genYAML(yamlconfig, output=output_file)
NameError: name 'genYAML' is not defined
```

Why does importing `genYAML` from `ufsda.genYAML` turn off info level logging?

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: The following `import` section is inside `def run_jedi_exe(yamlconfig):` in `run_jedi_exe.py`
```
    # add your ufsda python package to path
    ufsda_path = os.path.join(all_config_dict['GDASApp home'], 'ush')
    sys.path.append(ufsda_path)
    import ufsda
    from ufsda.misc_utils import calc_fcst_steps
    from ufsda.stage import gdas_single_cycle, gdas_fix, background_ens, atm_obs, bias_obs
```
I removed `from ufsda.genYAML import genYAM` from the top of `run_jedi_exe.py` and added it to the above `import` section.    With this change, `run_jedi_exe.py` runs to completion with info messages written to stdout.    This change was committed at [`f57fae0`](https://github.com/RussTreadon-NOAA/GDASApp/commit/f57fae0199ccdf2ff863623cb52ff9cb97f99210)

I don't understand why this change yields the desired behavior.

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: Is it possible it was using a different version of `genYAML` when it was imported before the `sys.path.append` section? I wonder if that has anything to do with it. Otherwise this is very strange indeed.

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Various tests this morning adding `print` to see what's loaded or moving `import` around did not uncover any explanation for the observed behavior.   `run_jedi_exe.py` at [`f57fae0`](https://github.com/RussTreadon-NOAA/GDASApp/commit/f57fae0199ccdf2ff863623cb52ff9cb97f99210) runs with log statements correctly written to stdout.

**Comment by RussTreadon-NOAA** [comment]
Rebuilt GDASApp 3/27 evening.  Much to my surprise 3dvar and 3denvar jobs launched by `run_jedi_exe.py` failed.    

The GDASApp build clones the head of gsibec develop.   
```
 # Build GSI-B
   option(BUILD_GSIBEC "Build GSI-B" OFF)
   if(BUILD_GSIBEC)
    ecbuild_bundle( PROJECT gsibec GIT "https://github.com/GEOS-ESM/GSIbec.git" BRANCH develop )
   endif()
```

I reached out to Ricardo Todling (GMAO).  He advised that we use the gsibec hash referenced in saber.   This is gsibec tag 1.0.7.   Update GDASApp top level `CMakeLists.txt` to clone
```
 # Build GSI-B
   option(BUILD_GSIBEC "Build GSI-B" OFF)
   if(BUILD_GSIBEC)
    ecbuild_bundle( PROJECT gsibec GIT "https://github.com/GEOS-ESM/GSIbec.git" TAG 1.0.7 )
   endif()
```
Rebuilt GDASApp.   The 3dvar and 3denvar tests successfully ran to completion with gsibec tag 1.0.7.   Commit this change `feature/3denvar` at 018cf8b.

**Comment by RussTreadon-NOAA** [comment]
**Testing**

- Automated GW-RT tests of `feature/3denvar` at [741fd8f](https://github.com/NOAA-EMC/GDASApp/pull/405/commits/741fd8fded8fc28532a8bfc5b9e2882d340d828a) pass on Hera and Orion.   
- Script `run_jedi_exe.py` successfully launches and runs 3dvar, 3denvar, 3dhofx, 4dhofx, and letkf on Hera and Orion.

The changes in this PR are ready for review.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA do we wish to review/merge this PR before we have fixed the genYAML issues? I'm fine with that approach but wanted to check with you before I proceed with reviewing this.

**Comment by RussTreadon-NOAA** [comment]
@CoryMartin-NOAA , if I had a quick fix for the issue documented in issue #408, I'd add it to this PR. 
 I don't.   Given this, I'm inclined to move forward with this PR.  Resolution of the problem noted in issue #408 can be the subject of a new PR.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA works for me, more agile!

File: test/atm/run_jedi_exe_3denvar.sh
Comment (CoryMartin-NOAA) [inline test/atm/run_jedi_exe_3denvar.sh]: @RussTreadon-NOAA do you want me to put the cases in the fix directory or would you like to be the case maintainer? I'm fine either way.

Comment (RussTreadon-NOAA) [reply test/atm/run_jedi_exe_3denvar.sh]: My preference is to keep things together in `gdasfix`.   Long term `gdasfix` could be in a role account directory so those in the group could share responsibility for maintaining `gdasfix`.

Comment (CoryMartin-NOAA) [reply test/atm/run_jedi_exe_3denvar.sh]: agreed, I'll copy over after lunch.

Comment (CoryMartin-NOAA) [reply test/atm/run_jedi_exe_3denvar.sh]: Now:
`cominges=${gdasfix}/cases`

Comment (RussTreadon-NOAA) [reply test/atm/run_jedi_exe_3denvar.sh]: Thank you @CoryMartin-NOAA .   Given this change the following have been updated to no longer reference `cominges` or `comin_ges`:

- `test/atm/run_jedi_exe_3denvar.sh`
- `test/atm/run_jedi_exe_letkf.sh`
- `ush/examples/run_jedi_exe/3denvar_hera.yaml`
- `ush/examples/run_jedi_exe/3denvar_orion.yaml`
- `ush/examples/run_jedi_exe/letkf_hera.yaml`
- `ush/examples/run_jedi_exe/letkf_orion.yaml`
- `ush/run_jedi_exe.py`

These changes were committed to `feature/3denvar` at [00a28d3](https://github.com/RussTreadon-NOAA/GDASApp/commit/00a28d3cc5cb1adaae267b1cd2a0b24bd1251b0c).

File: ush/ufsda/stage.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/stage.py]: @RussTreadon-NOAA I think I remember seeing some discussion about the gsibec fix file directory structure? 

what about going back to `if staticb_source in ['bump', 'gsibec']:` but making sure `case_anl` is consistent regardless of BUMP/GSIBEC? I would think we should be able to have independent `case` and `case_anl` if possible.

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Hmm, I'll need to think about this.   The current approach gets back to closed issue #403.

The way gsibec is implemented in UFS-DA we need the gsibec `case` to equal the background `case`.   The `case` gsibec may be used to generate a variational analysis on the `case_anl` grid.    For example, `fv3jedi_var.x` can read in `case=C96` tiles and then perform a gsibec 3dvar analysis on the `case_anl=C96` or `case_anl=C48` grid.   

The gsibec logic in `stage.py` is designed for this flexibility.   gsibec is associated with `case` while the analysis resolution is specified by `case_anl`.

We could refactor this section of `stage.py` so single `ufsda.disk_utils.symlink` line handles both bump and gsibec.   Would the following work & be preferable?

```
    # link staticb
    case_berror = case if staticb_source in ['gsibec'] else case_anl
    ufsda.disk_utils.symlink(os.path.join(input_fix_dir, staticb_source, case_berror),
                               config['fv3jedi_staticb_dir'])
```

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: yeah I think this is exactly what I had in mind, this approach reduces the amount of redundant code in the "if-block"

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Got it.  The suggested change was made and committed at 581bdd8.

File: ush/ufsda/stage.py
Comment (CoryMartin-NOAA) [inline ush/ufsda/stage.py]: Last comment before approval (sorry!)

So this section checks to see if the path doesn't exist, and then the try/except also does the `FileExistsError`. Is this just to be extra cautious in case it is created in this split second?
Or is this not doing what you want it to do? What about if the directory exists but it is not 'correct'? In this case it would not even get to the try/except and copy section.

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: @CoryMartin-NOAA , I confess to not putting much thought into this block.   I simply copied it from `background` in `stage.py` and placed it in a loop over the members.  I can't recall why I added the test `if not os.path.exists(ges_dir)`.   

This is the good thing about reviews.  They make you go back and think.  

Looking more closely I see that `background_ens` does exactly the same thing as the section of script above with the exception that `background_ens` creates member analysis directories.  This is necessary for enkf runs.   It's not necessary for EnVar runs.   

Instead of maintaining two blocks of script which basically do the same, I replaced the section of scripting discussed in this comment with `background_ens(config)`.  Logic has been added to `background_ens` to skip creation of member analysis directories when running EnVar.  

I did not commit this change yet because I'm not sure why we have `RESTART` and `RESTART_GES`   I don't see a need for `RESTART_GES`.   Do you recall why we have `RESTART_GES`?

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: The only thing that I can think of is if the analysis (RESTART) files are updated directly rather than passing increments. So then you keep the 'old' RESTART files as the guess and the new ones get updated and then used for the next forecast. We do this in the aerosol DA work right now because the model can't read the aerosol increments, but we can use the existing infrastructure for Gaussian increments after regridding for the ATM app, so no, I don't really see a point in having a `RESTART_GES` either.

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Using `background_ens` for EnVar and LETKF proved to be more involved than I anticipated.

`/parm/atm/lgetkf/lgetkf.yaml` includes `/RESTART` in the local path to each ensemble member.   `/parm/atm/berror/hybvar_gsibec.yaml` and  `parm/atm/berror/hybvar_bump.yaml` do not include `/RESTART` in the path.  I opted to remove `/RESTART` from `lgetkf.yaml`.   

There's another difference in the local path to the ensemble backgrounds for EnVar and LETKF.   The EnVar path use `ens`.   The LETKF path uses  `bkg`.   Logic was added to `background_ens` in `stage.py` to handle this.  

These changes were committed at 3904f75.

I currently have
```
    dohybvar = config['DOHYBVAR']
    bkgdir = 'ens'
    if not dohybvar: bkgdir = 'bkg'
```
in `background_ens`.   Is there a clever way to compress this to fewer lines?   My google searches have not yet turned up anything.

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: No I think that's about as compressed as it can get, unless you only use `dohybvar` once, then you can replace it with `config['DOHYBVAR']` on the last line and remove the first?

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Good suggestion.  Decide to directly use `config['DOHYBVAR']`.  Also clean up comments.  These changes committed at 854ed4b.

Manually reran 3dvar, 3denvar, letkf, 3dhofx, 4dhofx via `run_jedi_exe.py` on Orion.  All successfully run to completion.  

Manually ran 49 ctests within g-w `develop`.   All tests pass except one.  `test_gdasapp_atm_jjob_ens_run` failed.   Need to update g-w `scripts/exgdas_global_atmos_ensanal_run.sh` to account for removal of `RESTART` in ensemble member background path.   This g-w change will be addressed in g-w PR [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373).

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: Ok so since we have a chicken/egg problem and a queue of PRs in g-w, do we want to wait for that or merge this one and accept a failing test for the short term?

Comment (RussTreadon-NOAA) [reply ush/ufsda/stage.py]: Ideally we update g-w `develop` at the same time we merge this PR into GDASApp `develop`.   This won't happen.  

My preference is to see this PR enter GDASApp `develop` sooner rather than later.  I'm not keen on maintaining `feature/3denvar` until g-w is ready.  I've been maintaining several g-w branches for a few weeks waiting for PRs [#1370](https://github.com/NOAA-EMC/global-workflow/pull/1370), [#1372](https://github.com/NOAA-EMC/global-workflow/pull/1372), and [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373) to be reviewed and merged into g-w `develop`.

Comment (CoryMartin-NOAA) [reply ush/ufsda/stage.py]: Works for me, will do another review right now

File: ush/run_jedi_exe.py
Comment (CoryMartin-NOAA) [inline ush/run_jedi_exe.py:79]: just double checking you don't need `'atmos'` here any more in the path?

Comment (CoryMartin-NOAA) [reply ush/run_jedi_exe.py:79]: Never mind, I see below how it is used with the members

**Review by CoryMartin-NOAA — APPROVED** [review]

**Comment by CoryMartin-NOAA** [comment]
Thanks @RussTreadon-NOAA for all the hard work on this and the back and forth in review, much appreciated!

**Comment by CoryMartin-NOAA** [comment]
Will wait either for a 2nd approval or confirmation about the RT results being as expected before merging

**Comment by CoryMartin-NOAA** [comment]
These two single test failures are expected given earlier discussion with @RussTreadon-NOAA , I will merge this PR in and hope that the g-w PR can be merged relatively soon.

**Comment by RussTreadon-NOAA** [comment]
The Hera and Orion GW-RT failures of `test_gdasapp_atm_jjob_ens_run ` are expected.   

This PR removes `RESTART` from the run directory paths to the ensemble member backgrounds for ensemble DA.  This was done to be consistent with the run directory path to the deterministic backgrounds for variational DA.  The variational DA path does not include `RESTART`.  The run directory paths to the ensemble members for ensemble variational DA also do not include `RESTART`.

g-w `scripts/exgdas_global_atmos_ensanal_run.sh` needs to be updated to account for this change.  The required change will be made in g-w PR [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373).  Ideally g-w PR [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373) and GDASApp PR #405 would be merged into their respective `develop` branches around the same time.

**Comment by CoryMartin-NOAA** [comment]
Sorry @RussTreadon-NOAA if I jumped the gun on merging this. Let me try to review https://github.com/NOAA-EMC/global-workflow/pull/1373 later today and see if we can get the ball rolling on that.

**Comment by RussTreadon-NOAA** [comment]
Thank you @CoryMartin-NOAA  for merging this PR into GDASApp `develop`.   

It's OK to pause your review of g-w PR [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373).   g-w [#1370](https://github.com/NOAA-EMC/global-workflow/pull/1370) and [#1372](https://github.com/NOAA-EMC/global-workflow/pull/1372) are ahead of [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373).  EIB is currently reviewing PR [#1370](https://github.com/NOAA-EMC/global-workflow/pull/1370).

Now that GDASApp `develop` has been updated, I'll formally update g-w PR [#1373](https://github.com/NOAA-EMC/global-workflow/pull/1373).

**Pull request merged**
